docs/okeps/okep-5233-pre-assigned-pod-net-id.md

```diff
@@ -0,0 +1,344 @@
+
+// IPAM section contains IPAM-related configuration for the network.
+// +optional
+IPAM *IPAMConfig `json:"ipam,omitempty"`
+}
+
+// +kubebuilder:validation:XValidation:rule="isIP(self)", message="IP is invalid"
+type IP string
+
+// +kubebuilder:validation:MinItems=1
+// +kubebuilder:validation:MaxItems=2
+// +kubebuilder:validation:XValidation:rule="size(self) != 2 || !isIP(self[0]) || !isIP(self[1]) || ip(self[0]).family() != ip(self[1]).family()", message="When 2 IPs are set, they must be from different IP families"
+type DualStackIPs []IP
+
+```
+
+#### IPAMClaim API changes
+
+- TODO: Add status conditions to the IPAMClaim CRD, this will be an external enhancement in the network-plumbing group
+
+### Implementation Details
+
+#### Configurability
+
+The changes outlined in this enhancement should be configurable. This means a configuration knob
+is required to instruct OVN-Kubernetes on whether to process the annotation described in the 
+[Pod network identity](#pod-network-identity) section.
+
+#### NetworkSelectionElement annotation
+
+The `v1.multus-cni.io/default-network` annotation should only be processed for new pods, 
+modifying it after the addresses were allocated won't be reflected in the pods network 
+configuration and this should be blocked through a
+[Validating Admission Policy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/):
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: pre-assigned-network-identity
+spec:
+  matchConstraints:
+    resourceRules:
+      - apiGroups:   [""]
+        apiVersions: ["v1"]
+        operations:  ["UPDATE"]
+        resources:   ["pods"]
+  failurePolicy: Fail
+  validations:
+    - expression: "('v1.multus-cni.io/default-network' in oldObject.metadata.annotations) == ('v1.multus-cni.io/default-network' in object.metadata.annotations)"
+      message: "The 'v1.multus-cni.io/default-network' annotation cannot be changed after the pod was created"
+```
+The `NetworkSelectionElement` structure has an extensive list of fields, this enhancement 
+focuses only on the following:
+```cgo
+type NetworkSelectionElement struct {
+	// IPRequest contains an optional requested IP addresses for this network
+	// attachment
+	IPRequest []string `json:"ips,omitempty"`
+	// MacRequest contains an optional requested MAC address for this
+	// network attachment
+	MacRequest string `json:"mac,omitempty"`
+	// IPAMClaimReference container the IPAMClaim name where the IPs for this
+	// attachment will be located.
+	IPAMClaimReference string `json:"ipam-claim-reference,omitempty"`
+}
+```
+
+Any other field set in the struct will be ignored by OVN-Kubernetes.
+Both `Name` and `Namespace` are ignored as the primary UDN is known and cannot be changed.
+Note that `GatewayRequest` is not listed either, the default gateway is an attribute of the network is not going to be
+configurable per pod.
```

Comment:
Why don't we implement it by honor the NSE GatewayRequest ? what are the problems to do so ?

Reply:
that would only be useful for pods right ? @qinqon 

For VMs, we just need to ensure the `DHCPOptions` advertise the proper GW address (for IPv4) and make the VM forget the old gateway (by sending the RA from the configured IP address + lifetime=0). Right ?

Reply:
> that would only be useful for pods right ? @qinqon
> 

Yes.

> For VMs, we just need to ensure the `DHCPOptions` advertise the proper GW address (for IPv4) and make the VM forget the old gateway (by sending the RA from the configured IP address + lifetime=0). Right ?

An alternative is to configure the old gateway (global or LLA) at the gateway router Networks so VM has multipath but pointing to the same node.

Reply:
> Why don't we implement it by honor the NSE GatewayRequest ? what are the problems to do so ?

This enhancement covers setting the default gateway for the whole network, it doesn't cover setting custom gateways per workload. This is just not in scope an I wanted to call that out.